### ISSUE TITLE
Stream errors are not always handled or catchable

### ISSUE THREAD

**Issue by jgillman** [jimmynicol/image-resizer#46]
From my poking around I wasn't able to intercept any errors if there was a problem with the requested image. The most common situation is that an image is requested that doesn't exist and my logs are filling up with 500 errors :)

Personally, I would like to catch that error and return an appropriate default image.

I have two thoughts on how this could be accomplished:

1. This could be done inside of image-resizer with an optional env var that defines a path to a default image to be used if the requested path doesn't have a valid image.
2. Image-resizer could simply throw a catchable error and the user could add their own code to decide how to handle it. (e.g. serve a default image, throw some error code, etc.)

**Comment by jimmynicol** [comment]
@jgillman so any errors are passed down the stream and caught in the response stream. It would be fairly trivial to add a ENV option to render a default image on those types of errors. Idealistically this is a web service and returning 500s and no data is along the right lines, however practicalities suggests a solution here is in order.

I guess a good question to ask here is why are so many requests for images that dont exist being made? My usage for this service has in the past been only ever requesting images I already knew where there (like on Fundly.com) such there were very few 404 type requests being made.

**Comment by jgillman** [comment]
@jimmynicol We see very few 500s on production, I suspect most of them come from people hot-linking images off our servers and the image just doesn't exist any more.

This would almost exclusively be for our staging and development environments. Our production data is usually synced to our staging servers nightly and development boxes as needed, but data is always changing and the images aren't always where they're expected. By and large this is fine and it's a known "problem" in our company. It would be nice to serve up *something* in these cases where the actual image isn't as important as just getting *an* image in the correct dimensions.

Right now the route in our `index.js` file is:

```javascript
app.get('/*?', function(request, response){
  var image = new Img(request);

  image.getFile()
    .pipe(new streams.identify())
    .pipe(new streams.resizeSharp())
    .pipe(new streams.filter())
    .pipe(new streams.optimizeSharp())
    .pipe(streams.response(request, response));
});
```

Should I be able to listen on that last line as below?

```javascript
// ...
.pipe(streams.response(request, response).on( 'error', function() {
  // do something
});
//...

**Comment by jimmynicol** [comment]
@jgillman so you should actually attach your specific error handling directly to the stream you want to handle. As described [here](http://stackoverflow.com/questions/21771220/error-handling-with-node-js-streams) streams are not designed to forward errors. 

Your code would look something like this:

```javascript
image.getFile().on('error', function (err) {
  // handle error
}).pipe(new stream.identify())
...
```

So in cases where you are getting unfound objects from your S3 (or local) store you can throw a specific error and catch that to render a placeholder image.

**Comment by jgillman** [comment]
I've tried that, too. I'm not able to catch errors on `.getFile()` either.

```javascript
image.getFile()
  .on( 'error', function( error ) {
    console.log( 'there was an error!' + error );
  })
  .pipe(new streams.identify())
  // ...
```

The `console.log` never gets hit and I only see the usual output:

```
[resizer] new stream created!
[resizer] local filesystem - 2ms
[resizer] Error: ENOENT, open '/Users/jgillman/Code/local_images/fake-image-path.jpg'

GET /s200/local_images/uploads/merchant/main_image/125/mag-talll.jpg 500 19.794 ms - -
```

**Comment by jimmynicol** [comment]
@jgillman I just landed `v1.0.0` with the sharp engine (finally!). In addition to the new speed upgrades I also added `IMAGE_404` which allows you to specify an image to render if the app returns a 404 from any of the source scripts (s3, local, facebook etc).

In regards to your above error I realised I am catching all the errors and passing them into the stream so you wont be able to listen to any errors emitted from the various streams unless they are unhandled in some way. Sorry for missing that earlier.
